business/openid_auth.go

```diff
@@ -142,3 +326,142 @@ func GetOpenIdMetadata() (*OpenIdMetadata, error) {
+	if err != nil {
+		// On error, just inform that code flow is not possible
+		log.Warningf("Error when fetching OpenID provider's metadata: %s", err.Error())
+		return false
+	}
+
+	for _, v := range metadata.ResponseTypesSupported {
+		if v == "code" {
+			return true
+		}
+	}
+
+	return false
+}
+
+func RequestOpenIdToken(openIdParams *OpenIdCallbackParams, redirect_uri string) error {
+	openIdMetadata, err := GetOpenIdMetadata()
+	if err != nil {
+		return err
+	}
+
+	cfg := config.Get().Auth.OpenId
+
+	// Create HTTP client
+	httpTransport := &http.Transport{}
+	if cfg.InsecureSkipVerifyTLS {
+		httpTransport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	}
+
+	httpClient := http.Client{
+		Timeout:   time.Second * 10,
+		Transport: httpTransport,
+	}
+
+	// Exchange authentication code for a token
+	requestParams := url.Values{}
+	requestParams.Set("client_id", cfg.ClientId) // Can omit if not authenticated
+	requestParams.Set("code", openIdParams.Code)
+	requestParams.Set("grant_type", "authorization_code")
+	requestParams.Set("redirect_uri", redirect_uri)
+	response, err := httpClient.PostForm(openIdMetadata.TokenURL, requestParams)
+	if err != nil {
+		return fmt.Errorf("failure when requesting token from IdP: %w", err)
+	}
+
+	defer response.Body.Close()
+	rawTokenResponse, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read token response from IdP: %w", err)
+	}
+
+	if response.StatusCode != 200 {
+		return fmt.Errorf("request failed (HTTP response status = %s)", response.Status)
+	}
+
+	// Parse token response
+	var tokenResponse struct {
+		IdToken string `json:"id_token"`
+	}
+
+	err = json.Unmarshal(rawTokenResponse, &tokenResponse)
+	if err != nil {
+		return fmt.Errorf("cannot parse OpenId token response: %w", err)
+	}
+
+	if len(tokenResponse.IdToken) == 0 {
+		return errors.New("the IdP did not provide an id_token")
+	}
+
+	openIdParams.IdToken = tokenResponse.IdToken
+	return nil
+}
+
+func ValidateOpenIdNonceCode(openIdParams *OpenIdCallbackParams) (validationFailure string) {
+	// Parse the received id_token from the IdP and check nonce code
+	idTokenClaims := openIdParams.ParsedIdToken.Claims.(jwt.MapClaims)
+	nonceHashHex := fmt.Sprintf("%x", openIdParams.NonceHash)
+	if nonceClaim, ok := idTokenClaims["nonce"]; !ok || nonceHashHex != nonceClaim.(string) {
+		validationFailure = "nonce code mismatch"
+	}
+	return
+}
+
+// - CSRF mitigation
+func ValidateOpenIdState(openIdParams *OpenIdCallbackParams) (validationFailure string) {
+	state := openIdParams.State
+
+	separator := strings.LastIndexByte(state, '-')
+	if separator != -1 {
+		csrfToken, timestamp := state[:separator], state[separator+1:]
+		csrfHash := sha256.Sum224([]byte(fmt.Sprintf("%s+%s+%s", openIdParams.Nonce, timestamp, config.GetSigningKey())))
+
+		if fmt.Sprintf("%x", csrfHash) != csrfToken {
+			//RespondWithError(w, http.StatusForbidden, "Request rejected because of CSRF mitigation.")
```

Comment:
should we delete this unused/commented code?

Reply:
Hmm, yes, I'll delete.